stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -3,78 +3,216 @@
-     * @param y Second vector
-     * @return A vector of x and y concatenated together (in that order)
+     * return_type<int, int> promotes to double, which is not the desired
+     * behavior (hence this specialization)
+     *
+     * @tparam T1 First type to be promoted
+     * @tparam T2 Second type to be promoted
      */
-    inline std::vector<int>
-    append_array(const std::vector<int>& x, const std::vector<int>& y) {
-      std::vector<int> z;
-      z.reserve(x.size() + y.size());
-      z.insert(z.end(), x.begin(), x.end());
-      z.insert(z.end(), y.begin(), y.end());
-      return z;
-    }
+    template <>
+    struct append_return_type<int, int> {
+      typedef int type;
+    };
 
     /**
-     * Return the concatenation of two vectors containing Matrices
-     *   of the same type in the order of the arguments.
+     * Part of return type promotion logic for append_array
      *
-     * @tparam T1 Type of element in Matrices in first vector
-     * @tparam T2 Type of element in Matrices in second vector
-     * @tparam R Row specification of matrices
-     * @tparam C Column specification of matrices
-     * @param x First vector
-     * @param y Second vector
-     * @return A vector of x and y concatenated together (in that order)
+     * The type member is set to an Eigen Matrix with a scalar type that is the
+     * result of promoting the scalar types of the two provided Eigen Matrix
+     * types
+     *
+     * @tparam T1 Scalar type of first matrix type
+     * @tparam T2 Scalar type of first matrix type
+     * @tparam R Eigen RowsAtCompileTime of both matrices
+     * @tparam C Eigen ColsAtCompileTime of both matrices
      */
     template <typename T1, typename T2, int R, int C>
-    inline std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> >
-    append_array(const std::vector<Eigen::Matrix<T1, R, C> >& x,
-                 const std::vector<Eigen::Matrix<T2, R, C> >& y) {
-      if (!x.empty() && !y.empty()) {
-        check_matching_dims("append_array",
-                            "x[1]",
-                            x.front(),
-                            "y[1]",
-                            y.front());
+    struct append_return_type<Eigen::Matrix<T1, R, C>,
+                              Eigen::Matrix<T2, R, C> > {
+      typedef typename
+      Eigen::Matrix<typename append_return_type<T1, T2>::type, R, C> type;
+    };
+
+    /**
+     * Part of the return type promotion logic for append_array
+     *
+     * If the types of both template arguments are std::vectors, the type member
+     * is recursively computed as the append_return_type of the scalar types
+     * associated with those std::vectors.
+     */
+    template <typename T1, typename T2>
+    struct append_return_type<std::vector<T1>, std::vector<T2> > {
+      typedef typename
+      std::vector<typename append_return_type<T1, T2>::type> type;
+    };
+
+    /**
+     * Part of copy logic for append_array.
```

Comment:
Always say what the function does rather than how it fits into something larger.  The reasoning here is that the functions should themselves have coherent behavior that's easy to state so that they may be understood modularly.